lib/src/git.rs

```diff
@@ -757,3 +757,65 @@ pub struct Progress {
     pub bytes_downloaded: Option<u64>,
     pub overall: f32,
 }
+
+#[derive(Default)]
+struct PartialSubmoduleConfig {
+    path: Option<String>,
+    url: Option<String>,
+}
+
+/// Represents configuration from a submodule, e.g. in .gitmodules
+/// This doesn't include all possible fields, only the ones we care about
+#[derive(Debug, PartialEq)]
+pub struct SubmoduleConfig {
+    pub name: String,
+    pub path: String,
+    pub url: String,
+}
+
+pub fn parse_gitmodules(
+    gitmodules_path: &Path,
+) -> Result<BTreeMap<String, SubmoduleConfig>, git2::Error> {
+    let mut partial_configs: BTreeMap<String, PartialSubmoduleConfig> = BTreeMap::new();
+    let git_config = git2::Config::open(&gitmodules_path)?;
+
+    let entries = git_config.entries(Some(r#"submodule\..+\."#))?;
+    entries.for_each(|entry| {
+        let (config_name, config_value) = match (entry.name(), entry.value()) {
+            // Only accept entries with both name and value.
+            (Some(name), Some(value)) => (name, value),
+            _ => return,
+        };
+
+        // config_name is of the form submodule.<name>.<variable>
+        let (submod_name, submod_var) = config_name
+            .strip_prefix("submodule.")
+            .unwrap()
+            .split_once(".")
+            .unwrap();
+
+        let map_entry = partial_configs.entry(submod_name.to_string()).or_default();
+
+        match submod_var {
+            "path" => map_entry.path = Some(config_value.to_string()),
+            "url" => map_entry.url = Some(config_value.to_string()),
+            _ => return,
+        };
+    })?;
+
+    let ret: BTreeMap<String, SubmoduleConfig> = partial_configs
```

Comment:
nit: I would drop the type annotation if the compiler can infer the type (I think it should be able to).

Feel free to also drop the type annotation on `partial_configs`. If you think that makes it less clear, you could instead add a comment saying "partially built config for each submodule name" or something like that, which has the advantage of better describing the key than "String" does. I'm fine without the comment, though.

Reply:
Looks like I can drop the annotation from `ret`. Impressive.

It seems like I can't drop it from `partial_configs`, which makes sense because the body of `entries.for_each` doesn't use the key/value types directly (it uses `.entry`).